Application/EdFi.Ods.Api/Middleware/EdFiApiAuthenticationMiddleware.cs

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System;
+using System.Threading.Tasks;
+using EdFi.Ods.Common.Security;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace EdFi.Ods.Api.Middleware
+{
+    public class EdFiApiAuthenticationMiddleware
+    {
+        private readonly IApiKeyContextProvider _apiKeyContextProvider;
+        private readonly RequestDelegate _next;
+
+        public EdFiApiAuthenticationMiddleware(RequestDelegate next, IAuthenticationSchemeProvider schemes,
+            IApiKeyContextProvider apiKeyContextProvider)
+        {
+            if (schemes != null)
+            {
+                _next = next ?? throw new ArgumentNullException(nameof(next));
+                _apiKeyContextProvider = apiKeyContextProvider;
+
+                Schemes = schemes;
+            }
+            else
+            {
+                throw new ArgumentNullException(nameof(schemes));
+            }
+        }
+
+        public IAuthenticationSchemeProvider Schemes { get; set; }
+
+        public async Task Invoke(HttpContext context)
+        {
+            context.Features.Set<IAuthenticationFeature>(
+                new AuthenticationFeature
+                {
+                    OriginalPath = context.Request.Path,
+                    OriginalPathBase = context.Request.PathBase
+                });
+
+            // Give any IAuthenticationRequestHandler schemes a chance to handle the request
+            var handlers = context.RequestServices.GetRequiredService<IAuthenticationHandlerProvider>();
+
+            foreach (var scheme in await Schemes.GetRequestHandlerSchemesAsync())
+            {
+                var handler = await handlers.GetHandlerAsync(context, scheme.Name) as IAuthenticationRequestHandler;
+
+                if (handler != null && await handler.HandleRequestAsync())
+                {
+                    return;
+                }
+            }
+
+            var defaultAuthenticate = await Schemes.GetDefaultAuthenticateSchemeAsync();
+
+            if (defaultAuthenticate != null)
+            {
+                var result = await context.AuthenticateAsync(defaultAuthenticate.Name);
+
+                if (result?.Principal != null)
+                {
+                    context.User = result.Principal;
+
+                    // NOTE: For our use case we set the api key context into the call context storage. The rest of this code
+                    // is the default implementation and may need to be update on version updates.
```

Comment:
Link to the source for future updates: https://github.com/dotnet/aspnetcore/blob/v3.1.9/src/Security/Authentication/Core/src/AuthenticationMiddleware.cs is what I'm assuming this class was sourced from?  We seem to be not quite identical as well.

Reply:
yes geoff decompiled the assembly

Reply:
We should match source then, we're both Apache so we can share the source code.